src/dist/dist.rs

```diff
@@ -599,6 +600,87 @@ fn update_from_dist_<'a>(
     profile: Option<Profile>,
     prefix: &InstallPrefix,
     force_update: bool,
+) -> Result<Option<String>> {
+    let mut toolchain = toolchain.clone();
+    let mut fetched = String::new();
+    let mut first_err = None;
+    let backtrack = toolchain.channel == "nightly" && toolchain.date.is_none();
+    loop {
+        match try_update_from_dist_(
+            download,
+            update_hash,
+            &toolchain,
+            profile,
+            prefix,
+            force_update,
+            &mut fetched,
+        ) {
+            Ok(v) => break Ok(v),
+            Err(e) => {
+                if !backtrack {
+                    break Err(e);
+                }
+
+                if let ErrorKind::RequestedComponentsUnavailable(components, ..) = e.kind() {
+                    (download.notify_handler)(Notification::SkippingNightlyMissingComponent(
+                        components,
+                    ));
+
+                    if first_err.is_none() {
+                        first_err = Some(e);
+                    }
+                } else if let Some(e) = first_err {
+                    // if we fail to find a suitable nightly, we abort the search and give the
+                    // original "components unavailable for download" error.
+                    break Err(e);
+                } else {
+                    break Err(e);
+                }
+
+                // The user asked to update their nightly, but the latest nightly does not have all
+                // the components that the user currently has installed. Let's try the previous
+                // nightly (recursively) until we find a nightly that does.
+                //
+                // NOTE: we don't need to explicitly check for the case where the next nightly to
+                // try is _older_ than the current nightly, since we know that the user's current
+                // nightlys supports the components they have installed, and thus would always
+                // terminate the search.
+                if let Some(ref mut date) = toolchain.date {
```

Comment:
My gut feeling is that this could potentially be neater done as something like the following (untested):

```rust
toolchain.date = Some(Utc.from_utc_date(
    &toolchain.date.unwrap_or_else(
        ||NaiveDate::parse_from_str(&fetched, "%Y-%m-%d").expect("Malformed manifest date"))
    )).pred().format("%Y-%m-%d").to_string());
```

Obviously better formatting would be needed, and I've not checked the borrows vs. moves in that.

Reply:
Yeah, it was split up this way from back when I hadn't added `fetched`. Simplified it even further now :)

Reply:
Yeah, it was split up this way from back when I hadn't added `fetched`. Simplified it even further now :)